awscli/customizations/dynamodb/lexer.py

```diff
@@ -0,0 +1,250 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from base64 import b64decode
+import binascii
+import json
+import re
+import string
+
+
+VALID_BASE64 = re.compile(r'[A-Za-z0-9+/=]+')
+
+
+class LexerError(ValueError):
+    pass
+
+
+class Lexer(object):
+    START_IDENTIFIER = set(string.ascii_letters + '_')
+    VALID_IDENTIFIER = set(string.ascii_letters + string.digits + '_')
+    WHITESPACE = set(' \t\n\r')
+    SIMPLE_TOKENS = {
+        '.': 'dot',
+        ',': 'comma',
+        ':': 'colon',
+        '(': 'lparen',
+        ')': 'rparen',
+        '{': 'lbrace',
+        '}': 'rbrace',
+        '[': 'lbracket',
+        ']': 'rbracket',
+        '=': 'eq',
+    }
+    DIGITS = set(string.digits)
+    INT_CHARS = set(string.digits + '-')
+    OPERATION_TOKENS = {
+        'and', 'between', 'in', 'or', 'not',
+    }
+
+    def tokenize(self, expression):
+        self._init_expression(expression)
+        while self._current is not None:
+            if self._current in self.SIMPLE_TOKENS:
+                yield {
+                    'type': self.SIMPLE_TOKENS[self._current],
+                    'value': self._current,
+                    'start': self._position,
+                    'end': self._position + 1,
+                }
+                self._next()
+            elif self._current in self.START_IDENTIFIER:
+                yield self._consume_unquoted_identifier()
+            elif self._current == "'":
+                yield self._consume_quoted_identifier()
+            elif self._current == '"':
+                yield self._consume_string_literal()
+            elif self._current in self.WHITESPACE:
+                yield self._consume_whitespace()
+            elif self._current in self.INT_CHARS:
+                yield self._consume_number()
+            elif self._current == '<':
+                if self._next() == '>':
```

Comment:
It would be nice to encapsulate this logic into its own internal method as well.